third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1047,11 +1305,37 @@ std::shared_ptr<ReductionParams> getPersistentHeuristics(
   // The persistent buffer is detected as being t1, which would save the
   // persistent buffer as a float, however we could obviously just save t0 which
   // is half and would take half the memory. A more complex scenario of this
-  // which requires more advanced analysis is batch norm backwards.
+  // which requires more advanced analysis is inner_batch norm backwards.
   bool project_persistent_buffers =
       persistent_buffer_size_info.projected_persistent_buffer_size <
       persistent_buffer_size_info.persistent_buffer_size;
 
+  // for layer_norm backward, enable project to input can reuse weight shared
+  // among different rows. Although it increased register usage and may lead to
+  // register spills, the overall performance is increased
+  if (combined_inner_outer_reduction && !project_persistent_buffers) {
+    int64_t outer_reduction_buffer_size =
+        normalization_scheduler_utils::partialReductionBufferSize(
+            outer_reduction_tvs, runtime_info);
+    int64_t total_buffer_size =
+        persistent_buffer_size_info.projected_persistent_buffer_size +
+        outer_reduction_buffer_size;
+    // allow 10% more to allow project to input, 14K float should do project and
+    // 16K float should't do. more_register_factor >= 14*1024*5(three inputs,
+    // two outer reduciton results)*sizeof(float) /
+    // register_file_size_full
+    constexpr float more_register_factor = 1.1;
+    const int64_t avilable_register_file_size =
+        scheduler_utils::register_file_size_full * more_register_factor;
+    if (avilable_register_file_size >= total_buffer_size) {
+      project_persistent_buffers = true;
+      max_persistent_size = total_buffer_size;
+    } else {
+      max_persistent_size = persistent_buffer_size_info.persistent_buffer_size +
+          outer_reduction_buffer_size;
+    }
+  }
```

Comment:
Is this where the buffer size for the outer reduction is added to the persistent size?

Reply:
revised.
```
    // now we have the final decision on whether we project to input or not.
    if (project_persistent_buffers) {
      max_persistent_size =
          persistent_buffer_size_info.projected_persistent_buffer_size +
          outer_reduction_buffer_size;
    } else {
      max_persistent_size = persistent_buffer_size_info.persistent_buffer_size +
          outer_reduction_buffer_size;
    }
```